### PR TITLE
SPME improvements

### PR DESCRIPTION
**Pull request by rhalver**
Added LinkedCell usage for the computation of the Ewald sum to the SPME solver, fixing a bug
in the old implementation and replaced the computation of the auto-tuned parameters with
the Ewald parameter computation.

Now results of EwaldSum and SPME are identical for the potential, still no force computation.

(Merged ECP_Copa::Master)

### CODE REVIEW

**Comment by sfogerty** [comment]
@rhalver I sorted through some warnings of unused variables, and SPME now builds (warnings are treated as errors in our builds). Could you look at https://travis-ci.com/sfogerty/Cabana/jobs/226481124 and rearrange your initialization of variables there to clear up those warnings?

**Comment by rhalver** [comment]
> @rhalver I sorted through some warnings of unused variables, and SPME now builds (warnings are treated as errors in our builds). Could you look at https://travis-ci.com/sfogerty/Cabana/jobs/226481124 and rearrange your initialization of variables there to clear up those warnings?

Just pushed the changes for that, I hope I did not miss any. Otherwise I will fix them as well.

**Comment by sfogerty** [comment]
@rhalver thanks. Looks like a few re-orders and another unused variable needs a change. https://travis-ci.com/sfogerty/Cabana/jobs/226739818
(warnings start at line 1122)
I've looked at other things and the PR seems ready to accept once these annoying Travis errors are resolved

File: core/example/longrange/spme_example.cpp [hunk 1]
```diff
@@ -9,11 +9,12 @@ int main( int argc, char **argv )
...
 
     // crystal size
-    const int c_size = 2;
+    const int c_size = ( argc == 2 ) ? atoi( argv[1] ) : 32;
     // accuracy parameter for the tuning of Ewald
     const double accuracy = 1e-6;
     // width of unit cell (assume cube)
-    const double width = 1.0;
+    const double width = (double)c_size / 2.0;
+    ;
```
Comment (sfogerty) [inline core/example/longrange/spme_example.cpp]: Is this ";" a typo?

File: core/example/longrange/ewald.cpp [hunk 2]
```diff
@@ -34,63 +36,34 @@ void TEwald::tune( double accuracy, ParticleList particles, double lx,
...
+    _k_max = pow( EXECUTION_TIME_RATIO_K_R, 1.0 / 6.0 ) * sqrt( p / PI ) *
+             pow( N, 1.0 / 6.0 ) / l * 2.0 * PI;
+    _r_max = pow( EXECUTION_TIME_RATIO_K_R, 1.0 / 6.0 ) * sqrt( p / PI ) /
+             pow( N, 1.0 / 6.0 ) * l;
+    _alpha = sqrt( p ) / _r_max;
+    _k_max = 2.0 * sqrt( p ) * _alpha;
+
+    _k_max_int[0] = ceil( _k_max );
+    _k_max_int[1] = ceil( _k_max );
+    _k_max_int[2] = ceil( _k_max );
```
Comment (sfogerty) [inline core/example/longrange/ewald.cpp]: As I understand, k_max_int is no longer needed in this new implementation?

File: core/example/longrange/ewald.cpp [hunk 3]
```diff
@@ -103,210 +76,428 @@ double TEwald::compute( ParticleList &particles, double lx, double ly,
...
     auto q = Cabana::slice<Charge>( particles );
     auto p = Cabana::slice<Potential>( particles );
+    auto f = Cabana::slice<Force>( particles );
+    auto v = Cabana::slice<Velocity>( particles );
+    auto i = Cabana::slice<Index>( particles );
 
     int n_max = particles.size();
 
-    auto init_p = KOKKOS_LAMBDA( const int idx ) { p( idx ) = 0.0; };
+    auto init_p = KOKKOS_LAMBDA( const int idx )
```
Comment (sfogerty) [inline core/example/longrange/ewald.cpp]: If we init more than p, maybe we should call it something else?

File: core/example/longrange/ewald.cpp [hunk 4]
```diff
@@ -103,210 +76,428 @@ double TEwald::compute( ParticleList &particles, double lx, double ly,
...
-        Kokkos::RangePolicy<ExecutionSpace>( 0, n_max ),
-        KOKKOS_LAMBDA( int idx, double &Ur_part ) {
-            double d[SPACE_DIM];
-            double k;
-            // For each particle with charge q, the real space contribution to
-            // energy is Ur_part = 0.5*q*SUM_i(q_i*erfc(alpha*dist)/dist) The
-            // sum is over all other particles in the cell and in neighboring
-            // images up to some real-space cutoff distance "r_max"
-            //
-            // Self-energy terms are included then corrected for later, with one
```
Comment (sfogerty) [inline core/example/longrange/ewald.cpp]: I'd like to keep some version of these detailed comments. Could some description be added in the new LinkedCellList implementation?

File: core/example/longrange/ewald.cpp [hunk 5]
```diff
@@ -103,210 +76,428 @@ double TEwald::compute( ParticleList &particles, double lx, double ly,
...
+                    int jny = ( jcidx / nneigdim ) % nneigdim - (int)cfactor;
+                    int jnz = jcidx / ( nneigdim * nneigdim ) - (int)cfactor;
+
+                    // compute global cell index
+                    int jc = ( iz + jnz + cdim( 2 ) ) % cdim( 2 ) +
+                             ( iy + jny + cdim( 1 ) ) % cdim( 1 ) * cdim( 2 ) +
+                             ( ix + jnx + cdim( 0 ) ) % cdim( 0 ) * cdim( 1 ) *
+                                 cdim( 2 );
+
+                    // get neighbor coodinates
```
Comment (sfogerty) [inline core/example/longrange/ewald.cpp]: Nice descriptive comments here. I'd like to see a bit more later in this loop

File: core/example/longrange/spme.cpp [hunk 6]
```diff
@@ -206,88 +182,330 @@ double TPME::compute( ParticleList &particles, ParticleList &mesh, double lx,
...
+                    Kokkos::atomic_add( &f( i, 1 ), fy_i );
+                    Kokkos::atomic_add( &f( i, 2 ), fz_i );
+                    contrib += 2.0 * Ur_i;
                 }
-            }
-            Ur_part *= 0.5;
-            p( idx ) += Ur_part;
-        },
-        Ur );
-    Kokkos::fence();
```
Comment (sfogerty) [inline core/example/longrange/spme.cpp]: I had been putting fences after parallel loops where I thought GPU execution may need to completely finish before moving on. I wonder if I was being too cautious here, or would a fence be a good idea?

File: core/example/longrange/ewald.cpp (hunk 2 above)
Comment (rhalver) [reply core/example/longrange/ewald.cpp]: That is correct, I removed the elements from the class.

File: core/example/longrange/spme_example.cpp (hunk 1 above)
Comment (rhalver) [reply core/example/longrange/spme_example.cpp]: It is and is fixed now.

File: core/example/longrange/ewald.cpp (hunk 3 above)
Comment (rhalver) [reply core/example/longrange/ewald.cpp]: Renamed to init_parameters (unless you have a better name?)

File: core/example/longrange/ewald.cpp (hunk 4 above)
Comment (rhalver) [reply core/example/longrange/ewald.cpp]: Wrote a short bit about it, at the beginning of the branch

**Comment by sfogerty** [comment]
@rhalver Could you push those changes you mentioned to this branch? (e.g. removing k_max_int)

File: core/example/longrange/ewald.cpp (hunk 5 above)
Comment (rhalver) [reply core/example/longrange/ewald.cpp]: Updated the comments for the loops for the interactions between surrounding boxes and local boxes.

File: core/example/longrange/spme.cpp (hunk 6 above)
Comment (rhalver) [reply core/example/longrange/spme.cpp]: I have added a fence after the computations of Ur_ii, which as far as I can see could overlap with the computation of Ur_ij.

**Comment by rhalver** [comment]
> @rhalver Could you push those changes you mentioned to this branch? (e.g. removing k_max_int)

Was intending to finish it yesterday, but got interrupted during the extension of the comments, now everything should be fixed and it should be able to compile with -Werror.

**Pull request merged**